src/backend/distributed/commands/trigger.c

```diff
@@ -203,45 +242,480 @@ get_relation_trigger_oid_compat(HeapTuple heapTuple)
+}
+
+
+/*
+ * GetAlterTriggerDependsTriggerName returns Value object for the trigger
+ * name that given AlterObjectDependsStmt is executed for.
+ */
+static Value *
+GetAlterTriggerDependsTriggerNameValue(AlterObjectDependsStmt *alterTriggerDependsStmt)
+{
+	List *triggerObjectNameList = (List *) alterTriggerDependsStmt->object;
+	Value *triggerNameValue = llast(triggerObjectNameList);
+	return triggerNameValue;
+}
+
+
+/*
+ * PreprocessDropTriggerStmt is called before a DROP TRIGGER command has been
+ * executed by standard process utility. This function errors out for
+ * unsupported commands or creates ddl job for supported DROP TRIGGER commands.
+ */
+List *
+PreprocessDropTriggerStmt(Node *node, const char *commandString)
+{
+	DropStmt *dropTriggerStmt = castNode(DropStmt, node);
+	Assert(dropTriggerStmt->removeType == OBJECT_TRIGGER);
+
+	RangeVar *relation = GetDropTriggerStmtRelation(dropTriggerStmt);
+
+	bool missingOk = true;
+	LOCKMODE lockMode = AccessExclusiveLock;
+	Oid relationId = RangeVarGetRelid(relation, lockMode, missingOk);
+
+	if (!OidIsValid(relationId))
+	{
+		/* let standard process utility to error out */
+		return NIL;
+	}
+
+	if (!IsCitusTable(relationId))
+	{
+		return NIL;
+	}
+
+	ErrorIfUnsupportedTriggerCommand((Node *) dropTriggerStmt, lockMode);
+
+	if (IsCitusLocalTable(relationId))
+	{
+		char *triggerName = NULL;
+		ExtractDropStmtTriggerAndRelationName(dropTriggerStmt, &triggerName, NULL);
+		return CitusLocalTableTriggerCommandDDLJob(relationId, triggerName,
+												   commandString);
+	}
+
+	return NIL;
+}
+
+
+/*
+ * DropTriggerEventExtendNames extends relation name and trigger name with
+ * shardId, and sets schema name in given DropStmt by recreating "objects"
+ * list.
+ */
+void
+DropTriggerEventExtendNames(DropStmt *dropTriggerStmt, char *schemaName, uint64 shardId)
+{
+	Assert(dropTriggerStmt->removeType == OBJECT_TRIGGER);
+
+	char *triggerName = NULL;
+	char *relationName = NULL;
+	ExtractDropStmtTriggerAndRelationName(dropTriggerStmt, &triggerName, &relationName);
+
+	AppendShardIdToName(&triggerName, shardId);
+	Value *triggerNameValue = makeString(triggerName);
+
+	AppendShardIdToName(&relationName, shardId);
+	Value *relationNameValue = makeString(relationName);
+
+	Value *schemaNameValue = makeString(pstrdup(schemaName));
+
+	List *shardTriggerNameList =
+		list_make3(schemaNameValue, relationNameValue, triggerNameValue);
+	dropTriggerStmt->objects = list_make1(shardTriggerNameList);
+}
+
+
+/*
+ * ExtractDropStmtTriggerAndRelationName extracts triggerName and relationName
+ * from given dropTriggerStmt if arguments are passed as non-null pointers.
+ */
+static void
+ExtractDropStmtTriggerAndRelationName(DropStmt *dropTriggerStmt, char **triggerName,
+									  char **relationName)
+{
+	List *targetObjectList = dropTriggerStmt->objects;
+
+	/* drop trigger commands can only drop one at a time */
+	Assert(list_length(targetObjectList) == 1);
```

Comment:
```c
static void
ExtractDropStmtTriggerAndRelationName(DropStmt *dropTriggerStmt, char **triggerName,
									  char **relationName)
{
List *targetObjectList = dropTriggerStmt->objects;
if (list_length(targetObjectList) != 1) {
..error
}

/*
 * triggerObjectNameList will be either: 
  * - [schename, relationname, triggername]
  * - [relationname, triggername]
 */
List *triggerObjectNameList = linitial(targetObjectList);

int triggerNameindex = list_length(triggerObjectNameList) -1;
int relationNameIndex = triggerNameIndex - 1;
*triggerName =(some operation) list_nth(triggerObjectNameList, triggerNameIndex);
*relationName = (some operation) list_nth(triggerObjectNameList, relationNameIndex);

}
```

I would suggest removing SplitLastPointerElement and having a version that is similar to the above code snippet. 

I am a bit unsure how reliable it is to depend on postgres internal representation for trigger object names. But there doesn't seem to be an alternative ( at least currently we don't know). I think in this case we should be as robust as possible on our side. For example what happens if they add a new element to the list so that the last element is no longer the triggerName, then we could get a crash here.

How does the above sound?

Reply:
> List *targetObjectList = dropTriggerStmt->objects;
if (list_length(targetObjectList) != 1) {
..error
}

That can be the case, parser creates `targetObjectList` even if such a triggers does not exist.

> For example what happens if they add a new element to the list so that the last element is no longer the triggerName, then we could get a crash here.

This is the case for many points in citus like sequences (I can find many points in our codebase that relies on this fact): https://github.com/citusdata/citus/blob/bec58000d6cc7308e13dc78ec17746306c4bb17a/src/backend/distributed/commands/sequence.c#L138

I don't think postgres's internal data structures are changed frequently.

Reply:
I think having explicit `triggerNameindex ` and `relationNameIndex ` makes it easier to read/less prone to bugs

Reply:
But list length might be 2 or 3 depending on the case and we should use all the elements of the list except its last element to determine the relation (relation might be in a schema that is not in `search_path`)

Also, the way we parse triggerName & the relation is very identical to postgres. From postgres/dropcmds.c

```c
		case OBJECT_TRIGGER:
			if (!owningrel_does_not_exist_skipping(castNode(List, object), &msg, &name))
			{
				msg = gettext_noop("trigger \"%s\" for relation \"%s\" does not exist, skipping");
				name = strVal(llast(castNode(List, object)));
				args = NameListToString(list_truncate(list_copy(castNode(List, object)),
													  list_length(castNode(List, object)) - 1));
			}
```

Reply:
I guess that is the case in the other method, not in this one right? like with the call `makeRangeVarFromNameList`.

In the other method, we can get the schema name as well, and create a list with two elements, schemaName and relName.

Note: The reason I am saying this is to prevent future readability problems and make things more explicit, currently I am having a bit hard time understanding this part, so someone without a context could have similar problems as well.

Also one of the parts that increase readability in my opinion is:
 
```c
/*
 * triggerObjectNameList will be either: 
  * - [schename, relationname, triggername]
  * - [relationname, triggername]
 */
List *triggerObjectNameList = linitial(targetObjectList);
```

Because otherwise It is not certain what the list might have.

Other things are:
- I would expect `relationNameList` to have lists of relation names. Currently it either has `relationName` or `schemaName and relationName`, using the above could make it more explicit.
- What about we create a utility method that takes a trigger `DropStmt`, and fills 3 values, `schemaName`, `relationName`, and `triggerName`. If `schemaName` doesn't exist it could be set to NULL.  So we could modify `ExtractDropStmtTriggerAndRelationName` as `ExtractTriggerDropStmtObjectNames`, then we can call it from `GetDropTriggerStmtRelation` as well.

Reply:
> I guess that is the case in the other method, not in this one right?

This is right, we don't need to know schema name etc in  `ExtractDropStmtTriggerAndRelationName`

Looking into `makeRangeVarFromNameList`, actually `triggerObjectNameList ` can even have 4 objects (+`catalogname`).
But in the context of `DropTriggerEventExtendNames`,
* we don't need catalogname or schemaname
* list length may change depending on the situation.

So, for me, the only clean solution is to get the last element and then truncate and then get the last element as we know the order of the elements, instead of first checking list length and then parsing all the things.

> then we can call it from GetDropTriggerStmtRelation as well.

But for the `GetDropTriggerStmtRelation` we should use all the elements but the trigger name to determine the range var.

Reply:
Improved `ExtractDropStmtTriggerAndRelationName` & removed `SplitLastPointerElement`.

However, I should note that we still copy the namelist before creating the RangeVar but it shouldn't be a problem as we are just copying the list temporarily before creating the RangeVar, with the purpose of not modifying original namelist within the parsed utility command struct.